plugins/http_plugin/include/eosio/http_plugin/beast_http_session.hpp

```diff
@@ -0,0 +1,577 @@
+#pragma once
+
+#include <eosio/http_plugin/common.hpp>
+
+#include <algorithm>
+#include <cstdlib>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+#include <vector>
+#include <sstream>
+
+extern fc::logger logger;
+
+namespace eosio { 
+
+    using std::chrono::steady_clock;
+
+// Boost 1.70 introduced a breaking change that causes problems with construction of strand objects from tcp_socket
+// this is suggested fix OK'd Beast author (V. Falco) to handle both versions gracefully
+// see https://stackoverflow.com/questions/58453017/boost-asio-tcp-socket-1-70-not-backward-compatible
+#if BOOST_VERSION < 107000                
+    typedef tcp::socket tcp_socket_t;
+#else
+    typedef asio::basic_stream_socket<asio::ip::tcp, asio::io_context::executor_type> tcp_socket_t;
+#endif
+
+    using boost::asio::local::stream_protocol;
+    
+#if BOOST_VERSION < 107000
+    using local_stream = boost::asio::basic_stream_socket<stream_protocol>;
+#else 
+#if BOOST_VERSION < 107300
+    using local_stream = beast::basic_stream<
+    stream_protocol,
+    asio::executor,
+    beast::unlimited_rate_policy>;
+#else      
+using local_stream = beast::basic_stream<
+    stream_protocol,
+    asio::any_io_executor,
+    beast::unlimited_rate_policy>;    
+#endif    
+#endif
+
+    //------------------------------------------------------------------------------
+    // fail() 
+    //  this function is generally reserved in the case of a severe error which results 
+    //  in immediate termiantion of the session, with no response sent back to the client
+    //  currently also includes SSL "short read" error for security reasons:
+    // 
+    //  https://github.com/boostorg/beast/issues/38
+    //  https://security.stackexchange.com/questions/91435/how-to-handle-a-malicious-ssl-tls-shutdown
+    void fail(beast::error_code ec, char const* what)
+    {
+        fc_elog(logger, "${w}: ${m}", ("w", what)("m", ec.message()));
+        fc_elog(logger, "closing connection");
```

Comment:
Is it true that in all cases this is called it closes that connection?

Reply:
Within beast_http_session, yes.   But this function is  also called in beast_http_listener before a connection with the client has been established (failures of bind, etc.),  perhaps it can additionally accept an a string which indicates the intended outcome, whether it will be connection/session termination in the case of beast_http_session, or a failure to establish the listener on the port.